Homework #2

danilkolikov opened the pull request:
Коликов Д.А., группа М3239

Commit:
CityCam

Коликов Д.А., группа М3239

dtrounine commented on app/src/main/java/ru/ifmo/android_2015/citycam/CityCamActivity.java:
```diff
@@ -21,30 +34,168 @@
+        activity.errorInfo.setText("Вебкамеры не найдены");
+    }
+
+    private class DownloadTask extends AsyncTask<Void, Void, Result> {
```
Класс должен быть объявлен static, иначе он держит неявную ссылку на объект внешнего класса CityCamActivity

dtrounine commented on app/src/main/java/ru/ifmo/android_2015/citycam/model/Util.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    public static WebcamInfo parseRespond(Reader in) throws IOException {
+        JsonReader json = new JsonReader(in);
```
Здесь вы жёстко закладываетесь на формат json-а -- на порядок и количество полей. Это плохо, потому что формат может измениться. Обычно никто не гарантирует такой строгий формат. Обычно делают так:

json.beginObject();
while (json.hasNext()) {
    String name = json.nextName();
    switch(name) {
        case field1: // парсим нужное нам поле №1
        case field2: // парсим нужное нам поле №2
        defaut: json.skipValue(); break; // все остальное скипаем
    }
}
json.endObject();`

dtrounine commented on app/src/main/java/ru/ifmo/android_2015/citycam/CityCamActivity.java:
```diff
@@ -21,30 +34,168 @@
+        protected Result doInBackground(Void... params) {
+            Log.i(TAG, "Task started");
+            try {
+                HttpURLConnection conn = (HttpURLConnection)
```
connection и InputStream надо обязательно закрывать только в блоке finally -- иначе возможны утечки в случае ошибок

Commit:
Fixed 2nd homework

AsyncTask made static
Now classes are more error-safe
JSON parser made more universal

danilkolikov commented:
Исправил

dtrounine commented:
Принято
